pandas/tests/test_strings.py

```diff
@@ -635,13 +634,248 @@ def check_index(index):
+            ("single", "Dave", 0),
+            ("single", "Toby", 0),
+            ("single", "Maude", 0),
+            ("multiple", "robAndSteve", 0),
+            ("multiple", "robAndSteve", 1),
+            ("multiple", "abcdef", 0),
+            ("multiple", "abcdef", 1),
+            ("multiple", "abcdef", 2),
+        ], names=(None, None, "match"))
+        expected_df = DataFrame(
+            expected_tuples, expected_index, expected_columns)
+        computed_df = Si.str.extractall(named_pattern, re.VERBOSE)
+        tm.assert_frame_equal(computed_df, expected_df)
+
+        # MultiIndexed subject with names.
+        Sn = Series(subject_list, series_index)
+        Sn.index.names = ("matches", "description")
+        expected_index.names = ("matches", "description", "match")
+        expected_df = DataFrame(
+            expected_tuples, expected_index, expected_columns)
+        computed_df = Sn.str.extractall(named_pattern, re.VERBOSE)
+        tm.assert_frame_equal(computed_df, expected_df)
+        
+        # optional groups.
+        subject_list = ['', 'A1', '32']
+        named_pattern = '(?P<letter>[AB])?(?P<number>[123])'
+        computed_df = Series(subject_list).str.extractall(named_pattern)
+        expected_index = MultiIndex.from_tuples([
+            (1, 0),
+            (2, 0),
+            (2, 1),
+        ], names=(None, "match"))
+        expected_df = DataFrame([
+            ('A', '1'), 
+            (NA, '3'),
+            (NA, '2'),
+        ], expected_index, columns=['letter', 'number'])
+        tm.assert_frame_equal(computed_df, expected_df)
+
+        # only one of two groups has a name.
+        pattern = '([AB])?(?P<number>[123])'
+        computed_df = Series(subject_list).str.extractall(pattern)
+        expected_df = DataFrame([
+            ('A', '1'), 
+            (NA, '3'),
+            (NA, '2'),
+        ], expected_index, columns=[0, 'number'])
+        tm.assert_frame_equal(computed_df, expected_df)
+
+    def test_extractall_single_group(self):
+        # extractall(one named group) returns DataFrame with one named
+        # column.
+        s = Series(['a3', 'b3', 'd4c2'], name='series_name')
+        r = s.str.extractall(r'(?P<letter>[a-z])')
+        i = MultiIndex.from_tuples([
+            (0, 0),
+            (1, 0),
+            (2, 0),
+            (2, 1),
+        ], names=(None, "match"))
+        e = DataFrame({"letter":['a', 'b', 'd', 'c']}, i)
+        tm.assert_frame_equal(r, e)
+
+        # extractall(one un-named group) returns DataFrame with one
+        # un-named column.
+        r = s.str.extractall(r'([a-z])')
+        e = DataFrame(['a', 'b', 'd', 'c'], i)
+        tm.assert_frame_equal(r, e)
+
+    def test_extractall_no_matches(self):
+        s = Series(['a3', 'b3', 'd4c2'], name='series_name')
+        # one un-named group.
+        r = s.str.extractall('(z)')
+        e = DataFrame(columns=[0])
+        tm.assert_frame_equal(r, e)
```

Comment:
here are the tests that make sure extractall works even when there are no matches at all.